torch/distributed/fsdp/symbolic_trace.py

```diff
@@ -0,0 +1,198 @@
+                execution_info.module_execution_info_dict[execution_info.current_module].append(
+                    (execution_info.current_module, named_params)
+                )
+    elif kind == "call_module":
+        module = execution_info.current_module
+        named_params_list = list(module.named_parameters())
+        if named_params_list != []:
+            execution_info.module_execution_info_dict[module].append(
+                (module, named_params_list)
+            )
+        for (_, p) in named_params_list:
+            if p not in execution_info._traced_param_set:
+                execution_info.param_exec_order.append(p)
+                execution_info._traced_param_set.add(p)
+    return proxy
+
+
+def _patched_call_module(
+    call_module: Callable,
+    execution_info: _ExecutionInfo,
+    module: torch.nn.Module,
+    forward: Callable[..., Any],
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any]
+) -> Any:
+    """
+    Override of Tracer.call_module (see
+    https://pytorch.org/docs/stable/fx.html#torch.fx.Tracer.call_module).
+    Tracer.call_module is called in symbolic tracing for each non-root module.
+    This override intercepts the recording of each operation and
+    update execution_info.module_forward_order and execution_info.module_execution_info_dict.
+
+    Args:
+        call_module (Callable):
+            The call_module function to be patched.
+        execution_info (_ExecutionInfo):
+            Used to repord the execution information.
+        module, forward, args, kwargs: inputs to the call_module function.
+    """
+    # Update execution_info.module_forward_order
+    execution_info.module_forward_order.append(module)
+    # Update execution_info.module_execution_info_dict[current_module]
+    named_params_list = list(module.named_parameters())
+    if named_params_list != []:
+        execution_info.module_execution_info_dict[execution_info.current_module].append(
+            (module, list(module.named_parameters()))
+        )
+    # Stores away current_module for restoration later
+    old_current_module = execution_info.current_module
+    execution_info.current_module = module
+    # Initialize execution_info.module_execution_info_dict[module]. Note that if
+    # the forward of module is called multiple times, this will record
+    # the execution info of the last forward pass.
+    execution_info.module_execution_info_dict[module] = []
+    # Delegates into the normal Tracer.call_module method
+    output = call_module(module, forward, args, kwargs)
+    # Restores current_module
+    execution_info.current_module = old_current_module
+    return output
+
+
+def _patch_tracer(
+    tracer: torch.fx.Tracer,
+    root_module: torch.nn.Module
+) -> _ExecutionInfo:
+    """
+    Patches the input tracer so that during tracer.trace(), the forward order
+    of all modules and the parameter execution information are recorded.
+    root_module is the top-level module to be traced.
```

Comment:
It may be worthwhile to clarify if `root_module` should have any, some, or no FSDP wrapping. IIUC, from `_init_param_exec_order_wrap_policy()`, `root_module` should be the root module of the module hierarchy _without any_ FSDP wrapping.

Reply:
Done!